src/main/scala/mesosphere/marathon/core/storage/impl/BasePersistenceStore.scala

```diff
@@ -0,0 +1,141 @@
+
+  override def ids[Id, V]()(implicit ir: IdResolver[Id, V, Category, K]): Source[Id, NotUsed] = {
+    rawIds(ir.category).map(ir.fromStorageId)
+  }
+
+  protected def rawVersions(id: K): Source[OffsetDateTime, NotUsed]
+
+  final override def versions[Id, V](
+    id: Id)(implicit ir: IdResolver[Id, V, Category, K]): Source[OffsetDateTime, NotUsed] = {
+    rawVersions(ir.toStorageId(id, None))
+  }
+
+  protected def rawDelete(k: K, version: OffsetDateTime): Future[Done]
+
+  override def delete[Id, V](
+    k: Id,
+    version: OffsetDateTime)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    val storageId = ir.toStorageId(k, Some(version))
+    lockManager.executeSequentially(k.toString) {
+      rawDelete(ir.toStorageId(k, Some(version)), version)
+    }
+  }
+
+  protected def rawDeleteAll(k: K): Future[Done]
+
+  final override def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    val storageId = ir.toStorageId(k, None)
+    lockManager.executeSequentially(k.toString) {
+      rawDeleteAll(ir.toStorageId(k, None))
+    }
+  }
+
+  protected[storage] def rawGet(k: K): Future[Option[Serialized]]
+
+  override def get[Id, V](id: Id)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = async {
+    val storageId = ir.toStorageId(id, None)
+    await(rawGet(storageId)) match {
+      case Some(v) =>
+        Some(await(Unmarshal(v).to[V]))
+      case None =>
+        None
+    }
+  }
+
+  override def get[Id, V](id: Id, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = async {
+    val storageId = ir.toStorageId(id, Some(version))
+    await(rawGet(storageId)) match {
+      case Some(v) =>
+        Some(await(Unmarshal(v).to[V]))
+      case None =>
+        None
+    }
+  }
+
+  protected def deleteOld(k: K, maxVersions: Int): Future[Done] = async {
+    val versions = await(rawVersions(k).toMat(Sink.seq)(Keep.right).run()).sortBy(_.toEpochSecond)
+    // we always store the current version (twice), once as a versioned node and once as the current one.
+    val numToDelete = versions.size - maxVersions - 1
+    if (numToDelete > 0) {
+      val deletes = versions.take(numToDelete).map(v => rawDelete(k, v))
+      await(Future.sequence(deletes))
+    }
+    Done
+  }
+
+  protected def rawStore[V](k: K, v: Serialized): Future[Done]
+
+  override def store[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Done] = {
+    val storageId = ir.toStorageId(id, None)
+    lockManager.executeSequentially(id.toString) {
+      async {
+        val serialized = await(Marshal(v).to[Serialized])
+        val (_, _) = (
+          await(rawStore(storageId, serialized)),
+          await(rawStore(ir.toStorageId(id, Some(ir.version(v))), serialized)))
```

Comment:
There are cases, where we don't want to store a versioned entry (e.g.DeploymentPlan).
Perhaps add a class level flag, which indicates, if versions should be written or not.
We would create a PersistentStore for DeploymentPlans without versions.